nui/src/main/java/org/terasology/nui/widgets/types/math/RectanglefWidgetFactory.java

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+package org.terasology.nui.widgets.types.math;
+
+import org.terasology.joml.geom.Rectanglef;
+import org.terasology.nui.databinding.Binding;
+import org.terasology.nui.util.RectUtility;
+import org.terasology.nui.widgets.types.RegisterTypeWidgetFactory;
+import org.terasology.nui.widgets.types.TypeWidgetBuilder;
+import org.terasology.nui.widgets.types.TypeWidgetFactory;
+import org.terasology.nui.widgets.types.TypeWidgetLibrary;
+import org.terasology.reflection.TypeInfo;
+
+import java.util.Optional;
+
+@RegisterTypeWidgetFactory
+public class RectanglefWidgetFactory implements TypeWidgetFactory {
+    @Override
+    public <T> Optional<TypeWidgetBuilder<T>> create(TypeInfo<T> type, TypeWidgetLibrary library) {
+        if (!Rectanglef.class.equals(type.getRawType())) {
+            return Optional.empty();
+        }
+
+        TypeWidgetBuilder<Rectanglef> builder =
+            new RectanglefWidgetBuilder(library)
+                .add("x",
+                    rectBinding -> new Binding<Float>() {
+                        @Override
+                        public Float get() {
+                            return rectBinding.get().minX;
+                        }
+
+                        @Override
+                        public void set(Float value) {
+                            Rectanglef old = rectBinding.get();
+                            rectBinding.set(RectUtility.createFromMinAndSize(value, old.minY, old.maxX - old.minX, old.maxY - old.minY));
```

Comment:
**request:** Do we have to create new instances of the Rectangle here (for the binding to fire), or could we set the value of the rectangle contained in `rectBinding`? Maybe at least do something like this (no new instance creation)

```suggestion
                            old.minX = value;
                            rectBinding.set(old);
```

or this (new instance, but less manual computation).

```suggestion
                            rectBinding.set(new Rectanglef(value, old.minY).setSize(old.getSizeX(), old.getSizeY()));
```